### PR TITLE
chore(cfnspec): consume CloudFormation specification in parts

### PR DESCRIPTION
**Pull request by rix0rrr**
Whenever there are errors in the CloudFormation specification, we
currently have to fail the build and can't consume anything.

To ensure we make some progress, apply the following strategy instead:

- Split the spec into fragments, on a per-service basis.
- Consume those per-service spec updates that are valid; if updates
  are invalid, we will leave them at the old version.

This will produce an always-building spec, of which certain parts
may be outdated. Report the outdated parts in the CHANGELOG.

Notifying the CloudFormation team about spec errors is an out-of-band
process, and out of scope of this PR.

As a side effect of this work, formalize the spec manipulation we do
with JSON and patch files into a mini-standard called "JSON Patch
Stacks", and add some tools to operate on them.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/cfnspec/build-tools/patch-stack.ts [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    throw new Error('Usage: patch-stack <DIR> [<FILE>]');
+  }
+
+  const [dir, targetFile] = args;
+
+  if (targetFile) {
+    await resolveStack(targetFile, dir, { quiet });
+  } else {
+    printSorted(await loadStack(dir, { quiet }));
+  }
```
Comment (njlynch) [inline packages/@aws-cdk/cfnspec/build-tools/patch-stack.ts:213]: Just from a logical flow perspective, I'd probably rewrite this something like:

```suggestion
  const stack = await loadStack(dir, { quiet });
  if (targetFile) {
    writeSorted(targetFile, stack);
  } else {
    printSorted(stack);
  }
```

.. and just get rid of `resolveStack` altogether.

File: packages/@aws-cdk/cfnspec/build-tools/patch-stack.ts [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+ * Stacks are loaded individually, and merged afterwards.
+ */
+export async function resolveStack(targetFile: string, sourceDirectory: string, options: PatchOptions = {}) {
+  const stack = await loadStack(sourceDirectory, options);
+  await writeSorted(targetFile, stack);
+}
+
+export async function writeSorted(targetFile: string, data: any) {
+  // eslint-disable-next-line @typescript-eslint/no-require-imports
+  const sortJson = require('sort-json');
```
Comment (njlynch) [inline packages/@aws-cdk/cfnspec/build-tools/patch-stack.ts:68]: Any particular reason for this inlined require (both here and in printSorted)? Is there an implementation reason, or is it just stylistic?

File: packages/@aws-cdk/cfnspec/build-tools/patch-stack.ts [hunk 3]
```diff
@@ -0,0 +1,232 @@
+
+/**
+ * Apply a JSON patch stack into the given target file
```
Comment (njlynch) [inline packages/@aws-cdk/cfnspec/build-tools/patch-stack.ts:3]: `~` I would have gone with `patch chain` or `patch sequence` or something like that. `Stack` is such an overloaded term in the CDK that it's hard to read `patch-stack` and not be thinking about CloudFormation stacks (or stack sets, or CDK stacks, or ...).

File: packages/@aws-cdk/cfnspec/spec-source/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
+# Specification source
+
+This directory contains specifications for the following things:
+
+- `specifictation`: CloudFormation and SAM resource specification
```
Comment (njlynch) [inline packages/@aws-cdk/cfnspec/spec-source/README.md:5]:
```suggestion
- `specification`: CloudFormation and SAM resource specification
```

File: packages/@aws-cdk/cfnspec/spec-source/README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+- `cfn-lint`: information that is sourced from the
+  [cfn-lint](https://github.com/aws-cloudformation/cfn-lint) project. This
+  directory is treated as a collection of plain JSON files and patches just like
+  the root.
+- `cfn-docs`: information that is sourced from the CloudFormation documentation website.
+  Contains the documentation strings for all CloudFormation resources and properties.
+
+Each of these directories contains a "JSON patch stack".
```
Comment (njlynch) [inline packages/@aws-cdk/cfnspec/spec-source/README.md:13]: `~` Chain/Sequence/Queue :)
```suggestion
Each of these directories contains a "JSON patch chain".
```

File: packages/@aws-cdk/cfnspec/build-tools/split-cfn-to-stack.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+ */
+import * as path from 'path';
+import * as fs from 'fs-extra';
+
+import { writeSorted } from './patch-stack';
+import { CfnSpec, CfnSpecValidator, formatErrorInContext } from './validate-cfn';
+
+async function main(args: string[]) {
+  if (args.length < 2) {
+    throw new Error('Usage: split-spec-by-service <SPECFILE> <SCHEMA> <DIRECTORY>');
```
Comment (njlynch) [inline packages/@aws-cdk/cfnspec/build-tools/split-cfn-to-stack.ts:22]: `~` I actually like this name `split-spec-by-service` better than `split-cfn-to-stack`.

Usage doesn't appear to match implementation.

File: packages/@aws-cdk/cfnspec/build-tools/validate-cfn.ts [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+ * - instance.PropertyTypes["..."].Properties.Xyz requires property "PrimitiveType"
+ * - instance.PropertyTypes["..."].Properties.Xyz requires property "Type"
+ * - instance.PropertyTypes["..."].Properties.Xyz requires property "Type"
+ * - instance.PropertyTypes["..."].Properties.Xyz requires property "ItemType"
+ * - instance.PropertyTypes["..."].Properties.Xyz requires property "Type"
+ * - instance.PropertyTypes["..."].Properties.Xyz requires property "PrimitiveItemType"
+ * - instance.PropertyTypes["..."].Properties.Xyz is not exactly one from "Primitive","Complex Type","Collection of Primitives","Collection of Complex Types"
+ * ```
+ *
+ * No bueno.
```
Comment (njlynch) [inline packages/@aws-cdk/cfnspec/build-tools/validate-cfn.ts:23]: `~` Might be nice to instead (or also) show what this outputs, so we can see how much bueno-er this is (and why it's worth maintaining this extra ~300 line file).

**Review by njlynch — CHANGES_REQUESTED** [review]
Looks good!

This is a good bit of new code, but no new tests. :( Some baseline tests so someone modifying this in the future knows they likely haven't horked it would be nifty.

Also, the build is currently failing. :)

File: packages/@aws-cdk/cfnspec/build-tools/patch-stack.ts [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+  if (targetFile) {
+    await resolveStack(targetFile, dir, { quiet });
+  } else {
+    printSorted(await loadStack(dir, { quiet }));
+  }
+}
+
+function eatArg(arg: string, args: string[]) {
+  for (let i = 0; i < args.length && args[i] !== '--'; i++) {
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/cfnspec/build-tools/patch-stack.ts:217]: I didn't want to invest in figuring out how to invoke `yargs` properly again (kinda hate it tbh), so I skimped on the argument parsing here and put in a minimal version. Then I realized with this code you couldn't process a file called `-q`, so I put in the "end of options" argument that's customary.

But there's not really any point to, and I did it wrong anyway, so I'm taking it out.

File: packages/@aws-cdk/cfnspec/build-tools/patch-stack.ts (hunk 2 above)
Comment (rix0rrr) [reply packages/@aws-cdk/cfnspec/build-tools/patch-stack.ts:68]: It doesn't have a definition file so it needs to be the bad import. But yeah, I can lift it.

File: packages/@aws-cdk/cfnspec/build-tools/patch-set.ts [hunk 9]
```diff
@@ -90,7 +120,10 @@ function merge(target: any, fragment: any, jsonPath: string[]) {
         // eslint-disable-next-line max-len
         throw new Error(`Attempted to merge ${JSON.stringify(fragVal)} into incompatible ${JSON.stringify(specVal)} at path ${jsonPath.join('/')}/${key}`);
       }
-      if (typeof specVal !== 'object' && specVal !== fragVal) {
+      if (specVal == fragVal) {
```
Comment (njlynch) [inline packages/@aws-cdk/cfnspec/build-tools/patch-set.ts:123]: Intentional `==`, or intended to be `===`?

**Review by njlynch — APPROVED** [review]

**Pull request merged**